R/merge_sce_list.R

```diff
@@ -190,6 +259,23 @@ prepare_sce_for_merge <- function(sce,
   # Add `sce_name` to colnames so cell ids can be mapped to originating SCE
   colnames(sce) <- glue::glue("{sce_name}-{colnames(sce)}")
 
+  # merge library metadata
+  metadata_list <- metadata(sce)
+
+  # create a data frame with all the library specific information
+  library_metadata <- metadata_list[names(metadata_list) %in% combine_metadata_cols] |>
```

Comment:
After a bit of thought, while I think the idea of combining library metadata into a data frame like for samples is a noble one, I worry that it will be brittle. I think a simpler path might be to just combine _all_ of the metadata into a `library_metadata` list of lists. So much of this might reduce down to something like:

```
library_metadata <- list(metadata(sce)$library_id = metadata(sce)[names(metadata(sce)) != "sample_metadata"]
sample_metadata <- metadata(sce)[["sample_metadata"]]

metadata_list <- c(
    library_id = metadata(sce)[["library_id"]]
    sample_id =  metadata(sce)[["sample_id"]]
    library_metadata = library_metadata,
    sample_metadata = sample_metadata
  )

```

The main thing I am thinking here is that we would not have to maintain a list of `combine_metadata_cols` or deal with collapsing values where there might have been a list. We would, however, maintain the `library_id` and `sample_id` values, which should be combined into vectors (followed by `unique()`). If there are other values that should be maintained at the top level, we could do that as well. But I think maintaining a list of things we want to retain at that top level makes more sense than a list of things we want to combine.



One potential problem I see here is if there is dealing with the case of combining two already merged sce objects, which is probably something we want to account for (and add as a test case?). 

Also, there is probably other metadata that is specific to multi-sample libraries that we would want to maintain if merging merged libraries... I'm not quite sure what we would do there, but maybe dropping anything like that and recreating it would make sense anyway? (I'm thinking things like any integration algorithm data, which I assume we are storing).

Reply:
So part of the reason I didn't do this earlier is because the SCE objects don't seem to want to store lists of lists? So I have the following code in the function for preparing the individual objects prior to merging: 
```
# separate library and sample metadata
  metadata_list <- metadata(sce)
  library_metadata <- metadata_list[names(metadata_list) != "sample_metadata"]
  sample_metadata <- metadata_list$sample_metadata

  # combine into one list
  metadata_list <- c(
    library_id = metadata(sce)[["library_id"]],
    sample_id = metadata(sce)[["sample_id"]],
    library_metadata = library_metadata,
    sample_metadata = sample_metadata
  )
```
But then that gives a list of 26 elements instead of a list of 4. And then `names(metadata_list` gives: 

```
[1] "library_id"                               "sample_id"                                "library_metadata.library_id"             
 [4] "library_metadata.sample_id"               "library_metadata.salmon_version"          "library_metadata.reference_index"        
 [7] "library_metadata.total_reads"             "library_metadata.mapped_reads"            "library_metadata.mapping_tool"           
[10] "library_metadata.alevinfry_version"       "library_metadata.af_permit_type"          "library_metadata.af_resolution"          
[13] "library_metadata.usa_mode"                "library_metadata.af_num_cells"            "library_metadata.tech_version"           
[16] "library_metadata.transcript_type"         "library_metadata.include_unspliced"       "library_metadata.filtering_method"       
[19] "library_metadata.miQC_model"              "library_metadata.prob_compromised_cutoff" "library_metadata.scpca_filter_method"    
[22] "library_metadata.min_gene_cutoff"         "library_metadata.adt_scpca_filter_method" "library_metadata.normalization"          
[25] "library_metadata.adt_normalization"       "library_metadata.highly_variable_genes"  
```

So then if I combine objects, I get the same problem that I was trying to avoid which is a bunch of metadata entries in the list that are named the same way.

Reply:
You can ignore that comment... I just need to convert the `library_metadata` and `sample_metadata` vectors to lists before adding them.

Reply:
My bad for saying list and using `c()`! Should be:
```
metadata_list <- list(
    library_id = metadata(sce)[["library_id"]]
    sample_id =  metadata(sce)[["sample_id"]]
    library_metadata = library_metadata,
    sample_metadata = sample_metadata
  )
```